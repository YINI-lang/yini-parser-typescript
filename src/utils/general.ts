// export const debugPrint = (str: any = '') => {
//     const isDebug = !!process.env.IS_DEBUG
//     isDebug && console.debug('DEBUG: ' + str)
// }
