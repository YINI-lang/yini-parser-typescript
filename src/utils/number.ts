export const isNaNValue = (n: number): boolean => {
    return Number.isNaN(n)
}

export const isInfinityValue = (n: number): boolean => {
    return n === Infinity || n === -Infinity
}

export const isValidJSNumber = (str: string): boolean => {
    if (str.trim() === '') return false // Reject empty or whitespace-only.
    const n = Number(str)
    return !Number.isNaN(n) // True only if parse succeeded.
}
