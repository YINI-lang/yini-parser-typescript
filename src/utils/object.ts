type PlainObject = Record<string, unknown>

interface SortOptions {
    /** Sort nested objects too (default: false) */
    deep?: boolean
    /** Custom key comparator (default: a.localeCompare(b)) */
    comparator?: (a: string, b: string) => number
}

const isPlainObject = (value: unknown): value is PlainObject => {
    return (
        typeof value === 'object' &&
        value !== null &&
        !Array.isArray(value) &&
        Object.prototype.toString.call(value) === '[object Object]'
    )
}

/**
 * Returns a new object with keys sorted alphabetically.
 * Arrays and non-plain objects are preserved as-is (unless deep + they contain plain objects).
 */
export const sortObjectKeys = <T>(obj: T, options: SortOptions = {}): T => {
    const { deep = false, comparator } = options

    if (!isPlainObject(obj)) {
        // If it's not a plain object (array, Date, null, primitive), return as-is.
        return obj
    }

    const keys = Object.keys(obj).sort(
        comparator ?? ((a, b) => a.localeCompare(b)),
    )

    const sortedEntries = keys.map((k) => {
        const value = (obj as PlainObject)[k]
        if (deep) {
            if (isPlainObject(value)) {
                return [k, sortObjectKeys(value, options)]
            }
            if (Array.isArray(value)) {
                // Deep-sort any plain objects inside arrays, leave others as-is.
                return [
                    k,
                    value.map((item) =>
                        isPlainObject(item)
                            ? sortObjectKeys(item, options)
                            : item,
                    ),
                ]
            }
        }
        return [k, value]
    })

    // Preserve the original type T structurally.
    return Object.fromEntries(sortedEntries) as T
}

/**
 * Removes top-level properties with value `undefined`.
 * Does not recurse into nested objects or arrays.
 *
 * @param obj Any JS object
 * @returns A shallow copy with undefined properties removed,
 *          except in nested objects or arrays.
 *
 * @example
 * Input:
 * const input = {
 *     a: 1,
 *     b: undefined,
 *     c: { d: undefined, e: 2 },
 *     f: [1, undefined, 2]
 * };
 *
 * Output:
 *   {
 *     a: 1,
 *     c: { d: undefined, e: 2 },
 *     f: [1, undefined, 2]
 *   }
 */
export const removeUndefinedShallow = <T extends object>(
    obj: T,
): Partial<T> => {
    const cleaned: any = {}
    for (const [key, value] of Object.entries(obj)) {
        if (value !== undefined) {
            cleaned[key] = value
        }
    }
    return cleaned
}

/**
 * Recursively removes properties with value `undefined` from objects
 * and arrays.
 *
 * @param obj Any JS value (object, array, primitive)
 * @returns A deep-cloned copy with all `undefined` removed, including in
 *          nested objects and arrays.
 *
 * @example
 * Input:
 *   const input = {
 *       a: 1,
 *       b: undefined,
 *       c: {
 *           d: undefined,
 *           e: 2,
 *           f: [1, undefined, { g: undefined, h: 42 }]
 *       }
 *   };
 *
 * Output:
 *   {
 *     a: 1,
 *     c: {
 *         e: 2,
 *         f: [1, { h: 42 }]
 *     }
 *   }
 */
export const removeUndefinedDeep = <T>(obj: T): T => {
    if (obj === null || typeof obj !== 'object') {
        return obj // primitive value
    }

    if (Array.isArray(obj)) {
        // Clean each element, and filter out undefined entries
        return obj
            .map((item) => removeUndefinedDeep(item))
            .filter((item) => item !== undefined) as any
    }

    const cleaned: any = {}
    for (const [key, value] of Object.entries(obj)) {
        if (value !== undefined) {
            cleaned[key] = removeUndefinedDeep(value as any)
        }
    }
    return cleaned
}
