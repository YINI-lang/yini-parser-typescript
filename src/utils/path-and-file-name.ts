import path from 'path'

export const getFileNameExtension = (fullPath: string): string => {
    return path.extname(fullPath)
}
