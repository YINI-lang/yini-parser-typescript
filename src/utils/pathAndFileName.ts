/**
 * This file contains general path and file name helper functions (utils).
 * @note More specific YINI helper functions should go into yiniHelpers.ts-file.
 */
import path from 'path'

export const getFileNameExtension = (fullPath: string): string => {
    return path.extname(fullPath)
}
