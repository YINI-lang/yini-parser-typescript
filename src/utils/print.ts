/**
 * This file contains general system helper functions (utils).
 * @note More specific YINI helper functions should go into yiniHelpers.ts-file.
 */
import util from 'util'
import { isDebug, isDev, isProdEnv, isTestEnv } from '../config/env'

export const debugPrint = (str: any = '') => {
    isDebug() && console.debug('DEBUG: ' + str)
}

export const devPrint = (str: any = '') => {
    isDev() && !isTestEnv() && console.log('DEV: ' + str)
}

export const toPrettyJSON = (obj: any): string => {
    const str = JSON.stringify(obj, null, 4)
    return str
}

/** Pretty-prints a JavaScript object as formatted JSON to the console.
 * Strict JSON, all keys are enclosed in ", etc.
 */
export const printJSON = (obj: any, isForce = false) => {
    if (!isForce) {
        if (isProdEnv() || (isTestEnv() && !isDebug())) return
    }

    const str = toPrettyJSON(obj)
    console.log(str)
}

/**
 * Print a full JavaScript object in a human-readable way (not as JSON).
 * Not strict JSON, and shows functions, symbols, getters/setters, and class names.
 * @param isColors If true, the output is styled with ANSI color codes.
 */
export const printObject = (obj: any, isForce = false, isColors = true) => {
    if (!isForce) {
        if (isProdEnv() || (isTestEnv() && !isDebug())) return
    }

    console.log(util.inspect(obj, { depth: null, colors: isColors }))
}
