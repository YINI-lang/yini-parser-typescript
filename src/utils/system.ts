import { isDebug, isDev } from '../config/env'

export const debugPrint = (str: any = '') => {
    isDebug() && console.debug('DEBUG: ' + str)
}

export const devPrint = (str: any = '') => {
    isDev() && console.log('DEV: ' + str)
}

export const printObject = (obj: any) => {
    const str = JSON.stringify(obj, null, 4)
    console.log(str)
}
