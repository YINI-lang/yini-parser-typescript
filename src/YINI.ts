import fs from 'fs'
import { isDebug, isDev } from './config/env'
import { parseYINI } from './parseEntry'
import { getFileNameExtension } from './utils/path-and-file-name'
import { debugPrint, devPrint } from './utils/system'

export default class YINI {
    public static fullPath: string = '' // Used in error reporting.

    public static parse = (yiniContent: string): any => {
        debugPrint('-> Entered static parse(..) in class YINI\n')

        // Important: First, before anything, trim beginning and trailing whitespaces!
        yiniContent = yiniContent.trim()

        if (!yiniContent) {
            throw new Error('Syntax-Error: Unexpected blank YINI input')
        }
        if (!yiniContent.endsWith('\n')) {
            yiniContent += '\n'
        }

        debugPrint()
        debugPrint('==== Call parse ==========================')
        //const tree = parser.yini;  // Start rule.
        const result = parseYINI(yiniContent)
        debugPrint('==== End call parse ==========================\n')

        if (isDev()) {
            console.log()
            devPrint('YINI.parse(..): result:')
            console.log(result)
        }

        return result
    }

    /**
     *
     * @param yiniFile Full path to the YINI file.
     */
    public static parseFile = (fullPath: string): any => {
        debugPrint('Current directory = ' + process.cwd())

        if (getFileNameExtension(fullPath).toLowerCase() !== '.yini') {
            console.error('Invalid file extension for YINI file:')
            console.error(`"${fullPath}"`)
            console.log(
                'File does not have a valid ".yini" extension (case-insensitive).',
            )
            throw new Error('Error: Unexpected file extension for YINI file')
        }

        let content = fs.readFileSync(fullPath, 'utf8')
        let hasNoNewlineAtEOF = false

        if (!content.endsWith('\n')) {
            content += '\n'
            hasNoNewlineAtEOF = true
        }

        YINI.fullPath = fullPath
        const result: any = parseYINI(content)

        if (hasNoNewlineAtEOF) {
            console.warn(
                `No newline at end of file, it\'s recommended to end a file with a newline. File:\n"${fullPath}"`,
            )
        }

        if (isDev()) {
            console.log()
            devPrint('YINI.parseFile(..): result:')
            console.log(result)
        }

        return result
    }
}

// //import { Solution } from './solution';
// debugPrint('*** Started index.ts of ' + 'e_test'.toUpperCase() + ' ***')

// // const s : Solution = new Solution();

// // debugPrint('Result, getBuyDay index:  ', s.getBuyDay());
// // debugPrint('Result, getSellDay index: ', s.getSellDay());
// // debugPrint();

// const debugTestObj = {
//     name: 'e_test',
//     lang: 'TypeScript',
// }
// debugPrint('debugTestObj:')
// debugPrint(debugTestObj)
// debugPrint()

// const isDebug = !!process.env.IS_DEBUG
// debugPrint('env.IS_DEBUG:')
// debugPrint(process.env.IS_DEBUG)
// debugPrint('isDebug = ' + isDebug)
// debugPrint()

// const invalidInput = `
// # 	Config
// age = 30
// name = "Alice"
// items = ["a", "b", "c"]
// /END
// `

// const input1 = `
// ^ 	SectionName
// varBool = true
// varBool2 = off
// varInt = 30
// varFloat = 12.34
// varStr = "Alice"
// listItems = ["a", "b", "c"]
// varE1 = 1e4
// varE2 = 1.23e4
// varE3 = 6.5E23
// /END
// `

// const input2 = `
// ^ 	Config
// varAge = 30
// varName = "Alice"
// listItems = ["a", "b", "c"]
// 	^^Extra
// 	isExtra = true
// /END
// `

// // const input = `
// // # 	Config`;

// debugPrint('input1:')
// debugPrint(input1)
// debugPrint()

// YINI.parse(input2)
