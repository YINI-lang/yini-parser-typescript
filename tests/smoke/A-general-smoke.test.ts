import YINI from '../../src'
import { debugPrint } from '../../src/utils/system'

/**
 * General Smoke Tests.
 */
describe('General Smoke Tests:', () => {
    beforeAll(() => {})
    beforeAll(() => {
        debugPrint('beforeAll')

        // const isDebug = !!process.env.IS_DEBUG
        // if (!isDebug) {
        //     console.log('process.env.IS_DEBUG is false, OK')
        // } else {
        //     console.error('process.env.IS_DEBUG is true, FAIL')
        //     console.error(
        //         'Detected that IS_DEBUG is true, is should be false when testing',
        //     )
        //     console.error('process.env.IS_DEBUG:')
        //     console.error(process.env.IS_DEBUG)

        //     throw Error('ERROR: A variable in ENV has wrong state')
        // }
    })

    // Trivial test to see if everything is okey before moving on to "real" tests.
    test('1. Trivial test case.', () => {
        // Arrange.
        const a = 3
        const b = 5
        // Act.
        const result = a + b
        // Assert.
        expect(result).toEqual(8)
    })

    test('2. Short YINI Example.', () => {
        // Arrange.
        const validYini = `^ sectionTitle
            strVar = "Hello World!"
            intVar = 98.21`
        // Act.
        const result = YINI.parse(validYini)
        debugPrint(result)
        // Assert.
        expect(result.sectionTitle.strVar).toEqual('Hello World!')
        expect(result.sectionTitle.intVar).toEqual(98.21)
    })

    test('3. Should throw error if parsing some garbage.', () => {
        // Arrange.
        const fixture = `someGarbage`

        // Act & Assert.
        expect(() => {
            YINI.parse(fixture)
        }).toThrow()
    })

    test('4. Should throw error if parsing unknown file name.', () => {
        // Arrange.
        const fullPath = 'gibberish-path-and-file-name'

        // Act & Assert.
        expect(() => {
            debugPrint('fullPath = ' + fullPath)
            YINI.parseFile(fullPath)
        }).toThrow()
    })
})
