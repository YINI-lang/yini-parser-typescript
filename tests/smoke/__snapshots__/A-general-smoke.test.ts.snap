// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`General Smoke Tests: Sxxxxxx. 1`] = `
"Syntax-Error: Unexpected syntax while parsing a section head (section marker or section title):1
At line: 1, column(s): 1-1"
`;
